### PR TITLE
Flutter: 3.0.4 -> 3.3.3, flutter.dart: 2.17.5 -> 2.18.2, add a flutter2 package for "legacy" dependents

### PR DESCRIPTION
**Pull request by NANASHI0X74**
###### Description of changes
https://docs.flutter.dev/whats-new#aug-31-2022-flutter-vikings-edition-33-release

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NANASHI0X74** [comment]
Doesn't build currently:
```
$ nix-build /nix/store/90vj9b32w9q9dx63b3d6kmfbx85cmvlk-flutter-3.3.0-unwrapped.drv 
this derivation will be built:
  /nix/store/90vj9b32w9q9dx63b3d6kmfbx85cmvlk-flutter-3.3.0-unwrapped.drv
building '/nix/store/90vj9b32w9q9dx63b3d6kmfbx85cmvlk-flutter-3.3.0-unwrapped.drv'...
unpacking sources
unpacking source archive /nix/store/k0h6gcips67gj5r7ca1d7g99rdbbddcn-flutter_linux_3.3.0-stable.tar.xz
source root is flutter
setting SOURCE_DATE_EPOCH to timestamp 1661879220 of file flutter/.pub-cache/hosted/pub.dartlang.org/.cache/vector_math-versions.json
unpackPhase completed in 30 seconds
patching sources
applying patch /nix/store/2a5mrbr5scfvmbbjcfbn9ckac7bsyphz-disable-auto-update.patch
patching file bin/internal/shared.sh
patching file packages/flutter_tools/lib/src/runner/flutter_command_runner.dart
applying patch /nix/store/z2v0v71ba751ybmj3knf15lyybp13ny5-git-dir.patch
patching file dev/bots/prepare_package.dart
patching file packages/flutter_tools/lib/src/commands/downgrade.dart
patching file packages/flutter_tools/lib/src/version.dart
applying patch /nix/store/ym3mz067l0k2rri7vq6djjx72v1mklz8-move-cache.patch
patching file packages/flutter_tools/lib/src/asset.dart
patching file packages/flutter_tools/lib/src/cache.dart
patching file packages/flutter_tools/lib/src/artifacts.dart
patching file packages/flutter_tools/test/general.shard/artifacts_test.dart
patching script interpreter paths in ./bin/
./bin/cache/artifacts/gradle_wrapper/gradlew: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/1b9p07z77phvv2hf6gm9f28syp39f1ag-bash-5.1-p16/bin/bash"
./bin/flutter: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/1b9p07z77phvv2hf6gm9f28syp39f1ag-bash-5.1-p16/bin/bash"
./bin/dart: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/1b9p07z77phvv2hf6gm9f28syp39f1ag-bash-5.1-p16/bin/bash"
./bin/internal/update_dart_sdk.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/1b9p07z77phvv2hf6gm9f28syp39f1ag-bash-5.1-p16/bin/bash"
configuring
no configure script, doing nothing
building
/build/flutter/packages/flutter_tools /build/flutter
Resolving dependencies... 
  pedantic 1.11.1 (discontinued replaced by lints)
Got dependencies!
/build/flutter
Info: Compiling without sound null safety
Error: Error when reading 'packages/flutter_tools/.packages': No such file or directory
Error: Couldn't resolve the package 'flutter_tools' in 'package:flutter_tools/executable.dart'.
packages/flutter_tools/bin/flutter_tools.dart:7:8: Error: Not found: 'package:flutter_tools/executable.dart'
import 'package:flutter_tools/executable.dart' as executable;
       ^
packages/flutter_tools/bin/flutter_tools.dart:10:14: Error: Method not found: 'main'.
  executable.main(args);
             ^^^^
error: builder for '/nix/store/90vj9b32w9q9dx63b3d6kmfbx85cmvlk-flutter-3.3.0-unwrapped.drv' failed with exit code 254;
       last 10 log lines:
       > /build/flutter
       > Info: Compiling without sound null safety
       > Error: Error when reading 'packages/flutter_tools/.packages': No such file or directory
       > Error: Couldn't resolve the package 'flutter_tools' in 'package:flutter_tools/executable.dart'.
       > packages/flutter_tools/bin/flutter_tools.dart:7:8: Error: Not found: 'package:flutter_tools/executable.dart'
       > import 'package:flutter_tools/executable.dart' as executable;
       >        ^
       > packages/flutter_tools/bin/flutter_tools.dart:10:14: Error: Method not found: 'main'.
       >   executable.main(args);
       >              ^^^^
       For full logs, run 'nix log /nix/store/90vj9b32w9q9dx63b3d6kmfbx85cmvlk-flutter-3.3.0-unwrapped.drv'.
```

I wanted to debug this by running buildPhase in a nix-shell for flutter, but for some reason that doesn't work:
```
$ nix-shell /nix/store/90vj9b32w9q9dx63b3d6kmfbx85cmvlk-flutter-3.3.0-unwrapped.drv
warning: Nix search path entry '/nix/store/mvpahh0v6r5hmvnbbj386b855qmshxna-source/overlays' does not exist, ignoring

[nix-shell:~/Documents/projects/nixpkgs]$ cd flutter

[nix-shell:~/Documents/projects/nixpkgs/flutter]$ buildPhase 
```

I'll have to revisit this later, but help would be appreciated ^^

**Comment by NANASHI0X74** [comment]
In https://github.com/NixOS/nixpkgs/pull/189338/commits/f6f07a86ff9fd063b6c7d76199ca0fa113d3fdb2 I added a removed `import 'cache.dart'` back in after it was previously removed in the PR for flutter 3.0.0 because substituting `Cache.flutterRoot!` with `globals.fsUtils.homeDirPath!, '.cache', 'flutter',` doesn't work directly because it results in a too long argument list. Leaving it as is works for `flutter create` and `flutter run` though 😅

**Comment by NANASHI0X74** [comment]
I've tried this out with an existing app that's been updated to use the new flutter versions. flutter run, flutter pub get and dart's lsp server work without a hitch for me.
The builds for firmware-updater and fluffychat are broken again though:
```
error: builder for '/nix/store/xg95470ihsj0wbaaf8jvv60annxi3h7h-firmware-updater-unstable-deps-flutter-v3.3.2-x86_64-linux.tar.gz.drv' failed with exit code 1;
       last 10 log lines:
       > Downloading linux-x64/linux-x64-flutter-gtk tools...             2,420ms
       > Downloading linux-x64-profile/linux-x64-flutter-gtk tools...      1,512ms
       > Downloading linux-x64-release/linux-x64-flutter-gtk tools...      1,263ms
       >
       > 💪 Building with sound null safety 💪
       >
       > Building Linux application...                                           
       > CMake is required for Linux development.
       > It is likely available from your distribution (e.g.: apt install cmake), or can be downloaded from https://cmake.org/download/
       > mv: cannot stat '.packages': No such file or directory
       For full logs, run 'nix log /nix/store/xg95470ihsj0wbaaf8jvv60annxi3h7h-firmware-updater-unstable-deps-flutter-v3.3.2-x86_64-linux.tar.gz.drv'.
error: 1 dependencies of derivation '/nix/store/gpql8ygm4x4n9r8hwc40m3k5cg1c864b-firmware-updater-unstable.drv' failed to build
```

Should those packages use their own flutter versions to build and should a flutter_2 package be added? (https://github.com/NixOS/nixpkgs/pull/173200#issuecomment-1180170875)

**Comment by ericdallo** [comment]
> Should those packages use their own flutter versions to build and should a flutter_2 package be added?

IMO, yes

**Comment by SrTobi** [comment]
I use the changes here in my config and it's working great so far.

**Comment by riceicetea** [comment]
Hey!

I do definitely think that a `flutter_2` package should be added - this is the standard for a lot of packages (`python_x` from the top of my head).

Does this branch build currently? Do you have anything you need assistance with?

**Comment by NANASHI0X74** [comment]
flutter builds currently, some packages depending on it do not.
I know roughly what would need to be done for the flutter_2 package, just
haven't got round to it. I'd accept patches or PRs to my PR if you want to
do the work for me ^^

Sláinte,
Rian Lindenberger

**Comment by NANASHI0X74** [comment]
I took the liberty of refactoring flutter/default.nix a bit to make it simpler to update.I added a package flutter2 to build firmware-updater and fluffychat with, which works great. I tested that both of those applications run on my nixos machine. I just started another run of  `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` and will tick the box when it doesn't report any errors, otherwise you'll hear from me again.

**Review by ericdallo — APPROVED** [review]

File: pkgs/development/compilers/flutter/default.nix [hunk 1]
```diff
@@ -4,34 +4,40 @@ let
...
+      sources = {
+        "${dartVersion}-x86_64-linux" = fetchurl {
+          url = "https://storage.googleapis.com/dart-archive/channels/stable/release/${dartVersion}/sdk/dartsdk-linux-x64-release.zip";
+          sha256 = dartHash;
+        };
       };
     };
+    src = fetchurl {
+      url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
+      sha256 = hash;
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/flutter/default.nix:20]:
```suggestion
      inherit hash;
```

**Comment by Artturin** [comment]
i don't know  if anything in this pr caused this but discovered this when doing a nixpkgs-review on a another pr

```
error: hash mismatch in fixed-output derivation '/nix/store/p7xlzhqarn1h4i3y2adqg0g1m5axyr57-fluffychat-1.2.0-deps-flutter-v2.10.5-x86_64-linux.tar.gz.drv':
         specified: sha256-1PDX023WXRmRe/b1L+6Du91BvGwYNp3YATqYSQdPrRY=
            got:    sha256-b4BSOTnA3118HnzG8knLxFBu9NVn5puG/XesacySo10=
            ```

**Comment by NANASHI0X74** [comment]
@Artturin haven't looked into it, but just FYI the latest flutter update was https://github.com/NixOS/nixpkgs/pull/201564

this PR introduced a flutter2 package for older packages

**Comment by gilice** [comment]
> i don't know if anything in this pr caused this but discovered this when doing a nixpkgs-review on a another pr
> 
> ```
> error: hash mismatch in fixed-output derivation '/nix/store/p7xlzhqarn1h4i3y2adqg0g1m5axyr57-fluffychat-1.2.0-deps-flutter-v2.10.5-x86_64-linux.tar.gz.drv':
>          specified: sha256-1PDX023WXRmRe/b1L+6Du91BvGwYNp3YATqYSQdPrRY=
>             got:    sha256-b4BSOTnA3118HnzG8knLxFBu9NVn5puG/XesacySo10=
> ```

That's completely irrelevant. The fluffychat needs to be updated.

**Pull request merged**